experimental/PlaneCurve/EllCurveZnZ.jl

```diff
@@ -212,3 +213,534 @@ function rand_pair_EllCurve_Point(R::Oscar.MPolyRing_dec{S}, PP::Oscar.Geometry.
       error("Did not manage do find an elliptic curve with invertible discriminant")
    end
 end
+
+################################################################################
+# ECPP
+################################################################################
+# Cornacchia Algorithm
+################################################################################
+
+function evenodd(n::fmpz, fac::fmpz = ZZ(2))
+   res = (ZZ(1), n)
+   k = fac
+   g = gcd(k, n)
+   while g != ZZ(1)
+      if g < k
+         return (res[1]*g, div(res[2], g))
+      end
+      res = (res[1]*k, div(res[2], k))
+      k = k^2
+      g = gcd(k, res[2])
+   end
+   return res
+end
+
+################################################################################
+@doc Markdown.doc"""
+    cornacchia_algorithm(d::fmpz, m::fmpz)
+
+Return `true` and a solution of `x^2 + d*y^2 = m`if it exists, and false and
+`(0, 0)` otherwise.
+"""
+function cornacchia_algorithm(d::fmpz, m::fmpz)
+    (even, odd) = evenodd(m)
+    D = abs(d)
+    R = ResidueRing(ZZ, odd)
+    S = ResidueRing(ZZ, min(ZZ(8), even))
+    X = issquare_with_square_root(-R(D))
+    if !X[1]
+       return (false, (ZZ(0), ZZ(0)))
+    end
+    s1 = Hecke.data(X[2])
+    if !isone(even)
+       !isone(S(D)) && (false, (ZZ(0), ZZ(0)))
+       s2 = sqrtmod(-D, even)
+       s = abs(crt(-s1, odd, -s2+ZZ(2)^(Int(log2(even)-1)), even, true))
+    else
+       s = s1
+    end
+
+    if 2*s > m
+        B = m - s
+    else
+	    B = s
+    end
+
+    A = m
+    while B^2 >= m
+       dr = divrem(A, B)
+       A = B
+       B = dr[2]
+    end
+    t = m - B^2
+    L = divrem(t, D)
+    if !iszero(L[2])
+       return (false, (ZZ(0), ZZ(0)))
+    elseif !issquare(L[1])
+        return (false, (ZZ(0), ZZ(0)))
+    else
+        return (true, [B, isqrt(L[1])])
+    end
+end
+
+################################################################################
+###############Hilbert Class Polynomial#########################################
+################################################################################
+
+#Compute reduced forms of imaginary quadratic number fields
+# based on Algo 5.3.5 in the book by Cohen
+function class_data(D::fmpz)
+   D<ZZ(0) || error("Input needs to be negative")
+
+   h = 0
+   b = abs(D % 2)
+   sqrD3 = sqrt(div(BigInt(-D), 3))
+   B = BigInt(floor(sqrD3))
```

Comment:
what you want is `B = floor(fmpz, sqrD3)` which is works if `sqrD3` is a BigFloat. Even better
`B = root(div(-D, 3), 2)` does it all in one...